src/backend/distributed/planner/distributed_planner.c

```diff
@@ -347,6 +355,16 @@ ListContainsDistributedTableRTE(List *rangeTableList,
 			continue;
 		}
 
+		if (HideCitusDependentObjects && IsPgLocksTable(rangeTableEntry))
+		{
```

Comment:
I believe tidscan.sql shows the lock on `DistPartitionRelationId()`, is that right?

I wonder if we should `table_close(DistPartitionRelationId(), AccessShareLock)` if we are to `return false;` in this function?

In the end, if there are no distributed tables, we don't need to keep the lock. Would something like the below can avoid introducing `IsPgLocksTable`?

```
diff --git a/src/backend/distributed/metadata/metadata_cache.c b/src/backend/distributed/metadata/metadata_cache.c
index 681ca50ec..cb845af44 100644
--- a/src/backend/distributed/metadata/metadata_cache.c
+++ b/src/backend/distributed/metadata/metadata_cache.c
@@ -1447,7 +1447,7 @@ BuildCitusTableCacheEntry(Oid relationId)
        if (distPartitionTuple == NULL)
        {
                /* not a distributed table, done */
-               table_close(pgDistPartition, NoLock);
+               table_close(pgDistPartition, AccessShareLock);
                return NULL;
        }
 
diff --git a/src/backend/distributed/planner/distributed_planner.c b/src/backend/distributed/planner/distributed_planner.c
index 6d09e0023..98ebaca4d 100644
--- a/src/backend/distributed/planner/distributed_planner.c
+++ b/src/backend/distributed/planner/distributed_planner.c
@@ -355,16 +355,6 @@ ListContainsDistributedTableRTE(List *rangeTableList,
                        continue;
                }
 
-               if (HideCitusDependentObjects && IsPgLocksTable(rangeTableEntry))
-               {
-                       /*
-                        * Postgres tidscan.sql test fails if we do not filter pg_locks table because
-                        * test results, which show taken locks in serializable isolation mode,
-                        * fails by showing extra lock taken by IsCitusTable below.
-                        */
-                       continue;
-               }
-
                if (IsCitusTable(rangeTableEntry->relid))
                {
                        if (maybeHasForeignDistributedTable != NULL &&
@@ -375,6 +365,7 @@ ListContainsDistributedTableRTE(List *rangeTableList,
 
                        return true;
                }
```

Reply:
The failed query is

```sql
-- check predicate lock on CTID
BEGIN ISOLATION LEVEL SERIALIZABLE;
SELECT * FROM tidscan WHERE ctid = '(0,1)';
-- locktype should be 'tuple'
SELECT locktype, mode FROM pg_locks WHERE pid = pg_backend_pid() AND mode = 'SIReadLock';
ROLLBACK;
```

with the diff below:

```
 -- locktype should be 'tuple'
 SELECT locktype, mode FROM pg_locks WHERE pid = pg_backend_pid() AND mode = 'SIReadLock';
  locktype |    mode    
 ----------+------------
  tuple    | SIReadLock
-(1 row)
+ relation | SIReadLock
+(2 rows)
 
 ROLLBACK;
```

Reply:
Can we understand the whether we are running under serializable mode? Would be nice to add 
```C
if (HideCitusDependentObjects &&  InSerializableMode &&IsPgLocksTable(rangeTableEntry))
```
Such that this is more clear

Reply:
we can do that the macro IsolationIsSerializable() defined in server/access/xact.h